File: public/js/factories/comments_pre_handler.js
define(['factories', 'jquery', 'underscore'], function(factories, $, _) {
    return factories.factory('CommentsPreHandler', function() {
        return {
            render: function(text) {
                var id_comment_pair = {};
                if(!text.match(/\n$/)) text += '\n';
                text = text.replace(/<([^<>]+)>:\s*((.|\n)*?)(?=\n\s*(\n|<|$))/mg, function(match, id, comment) {
                    id_comment_pair[id] = comment;
                    return '';
                }).replace(/\[([^\[\]]+)\]\s*<([^<>]+)>/g, function(matched, content, id) {
                    if (id_comment_pair[id])
                        //TODO: rewrite
                        return compile(content.trim(), id_comment_pair[id].trim());
                    else
                        return matched;
                }).replace(/\[([^\[\]]+)\]{([^{}]+)}/g, function(matched, content, comments) {
                    return compile(content.trim(), comments.trim());
                });

                return text;

                function compile(content, comments) {
                    comments = _(comments.split("\n")).reduce(function(list, comment) {
                      var matches = comment.match(/(\w+):(.*)$/);
                      if (matches)
                        list.push({author: matches[1].trim(), comments: [matches[2].trim()]});
                      else if (_(list).isEmpty())
                        list.push({comments: [comment.trim()]});
                      else
                        _(list).last().comments.push(comment.trim());

                      return list;
                    }, []);

                    comments = _(comments).map(function(comment) {
                        var root = $('<div />').addClass('panel panel-default panel-comments');
                        if (comment.author)
                            root.append($('<div />').text(comment.author).addClass('panel-heading'));
                        if (!_(comment.comments).isEmpty()) {
                            var body = $('<div />').addClass('panel-body');
                            _(comment.comments).each(function(text) {
                                body.append($('<div />').text(text));
                            });
                            root.append(body);
                        }
                        return $('<div />').append(root).html();
                    });

                    var span = $('<span />').text(content).
                                 addClass('annotated').
                                 attr('data-toggle',    'popover').
                                 attr('data-placement', 'right').
                                 attr('data-animation', 'true').
                                 attr('data-html',      'true').
                                 attr('data-trigger',   'manual').
                                 attr('data-content',   comments.join(''));

                    return $('<div />').append(span).html();
                }
            }
        };
    });
});

File: lib/comments_pre_handler.rb
require 'nokogiri'

module CommentsPreHandler
  def render txt
    id_comment_pair = {}
    txt << "\n" unless txt.end_with? "\n" # Append newline automatically
    result =  txt.gsub(/\<([^\<\>]+)\>:\s*(.*?)(?=\n\s*(\n|\<|$))/m) do
                id_comment_pair[$1] = $2
                nil
              end.gsub(/\[([^\[\]]+)\]\s*\<([^\<\>]+)\>/) do |matched|
                content, id = $1, $2
                if id_comment_pair[id]
                  compile(content.strip, id_comment_pair[id].strip);
                else
                  matched
                end
              end.gsub(/\[([^\[\]]+)\]{([^{}]+)}/) do
                compile($1.strip, $2.strip)
              end
    super result
  end

  private
    def build &block
      html = Nokogiri::HTML::DocumentFragment.parse ''
      Nokogiri::HTML::Builder.with html, &block
      html.to_html
    end

    def compile content, comments
      comments = comments.split("\n").inject([]) do |list, comment|
        if comment =~ /(\w+):(.*)$/
          list << {author: $1.strip, comments: [$2.strip]}
        elsif list.empty?
          list << {comments: [comment.strip]}
        else
          list.last[:comments] << comment.strip
        end
        list
      end

      comments.map! do |comment|
        build do |doc|
          doc.div class: 'panel panel-default panel-comments' do
            doc.div comment[:author], class: 'panel-heading' if comment[:author]
            if comment[:comments].present?
              doc.div class: 'panel-body' do
                comment[:comments].each do |text|
                  doc.div text
                end
              end
            end
          end
        end.gsub("\n", '')
      end

      build do |doc|
        doc.span content,
        'class'          => 'annotated',
        'data-container' => 'body',
        'data-toggle'    => 'popover',
        'data-placement' => 'left',
        'data-animation' => 'true',
        'data-html'      => 'true',
        'data-trigger'   => 'manual',
        'data-content'   => comments.join
      end
    end
end
